File: python/p2.py
def fibonacci(lim):
	sum = 0
	idx = 0
	memo = [1, 1]
		
	while max(write_next(memo)) < lim:
		sum += (memo[idx]%2==0)*memo[idx]
		idx = (idx+1)%2
	return sum	

def write_next(pair):
	pair[pair.index(min(pair))] = sum(pair)
	return pair

# Sum of all even fibonacci numbers up to 4 million
print fibonacci(4*(10**6))

File: javascript/p2.js
const memo = [1, 1]

const fib = (n) => {
	if (memo[n]) return memo[n];

	m = memo.length;

	while (m < n) {
		memo.push(
			memo.slice(-2)
				.reduce((a, b) => { return a + b; })
		);
		m = memo.length;
	}
}

const fib_even_sum = (lim) => {
	while (memo.slice(-1) < lim) {
		fib(memo.length+1);
	}
	
	return memo.slice(0, -1)
				.filter((n) => { return !(n%2); })
				.reduce((a, b) => { return a + b; })
}

// Sum of all even fibonacci numbers up to 4 million
console.log(fib_even_sum(4*Math.pow(10, 6)))